### PR TITLE
std: stop using LinearFifo in BufferedReader

### PR DESCRIPTION
**Pull request by hwchen**
See also commit 7287c7482a2c694c7c7f56b9f7c1744a7ae7905f where LinearFifo is removed from BufferedWriter

Basically, LinearFifo does extra work v. copying bytes and increasing an offset.

---

### Context:

I wrote a clone of [runiq](https://github.com/whitfin/runiq), [zuniq](https://github.com/hwchen/zuniq), and was surprised `zuniq` was slower by quite a bit. On my test data set, runiq's read + filter + write had about the same time as just zuniq's read.

Just to give a broad idea of the speedup, here's some timings of before/after. (`patched` is the version run against the changes in this PR). I'm running ubuntu 20.04 on amd64 (ryzen).

```
zuniq on  main [+1] > just bench scratch/test-data.txt
zig build install -Drelease-fast -p zig-out
zig build install -Drelease-fast -p zig-out/patched --zig-lib-dir ~/zigdev/zig/lib
hyperfine --warmup 10 'zig-out/bin/zuniq scratch/test-data.txt' 'zig-out/patched/bin/zuniq scratch/test-data.txt' 'runiq scratch/test-data.txt'
Benchmark 1: zig-out/bin/zuniq scratch/test-data.txt
  Time (mean ± σ):     185.1 ms ±   4.2 ms    [User: 168.1 ms, System: 16.8 ms]
  Range (min … max):   179.1 ms … 191.6 ms    16 runs
 
Benchmark 2: zig-out/patched/bin/zuniq scratch/test-data.txt
  Time (mean ± σ):      84.7 ms ±   3.4 ms    [User: 66.6 ms, System: 17.9 ms]
  Range (min … max):    80.2 ms …  93.0 ms    36 runs
 
Benchmark 3: runiq scratch/test-data.txt
  Time (mean ± σ):     134.1 ms ±   3.9 ms    [User: 75.6 ms, System: 58.2 ms]
  Range (min … max):   126.1 ms … 140.9 ms    21 runs
 
Summary
  'zig-out/patched/bin/zuniq scratch/test-data.txt' ran
    1.58 ± 0.08 times faster than 'runiq scratch/test-data.txt'
    2.18 ± 0.10 times faster than 'zig-out/bin/zuniq scratch/test-data.txt'

zuniq on  main [+1] > ./zig-out/bin/zuniq scratch/test-data.txt| wc -l
522478
zuniq on  main [+1] > ./zig-out/patched/bin/zuniq scratch/test-data.txt| wc -l
522478
zuniq on  main [+1] > runiq scratch/test-data.txt| wc -l 
522478
```

---

### Todo

I've taken a shortcut in testing just as a sanity check: I used the latest nightly and ran `zig test --zig-lib-dir lib  --main-pkg-path lib/std lib/std/io/buffered_reader.zig`.

If this change is desirable, I'll take the time to:
- [x] Set up my zig dev environment properly to run the tests
- [x] Add some more tests, as I didn't check edge cases (which the LinearFifo may have tested)
- [x] Write a benchmark as in https://github.com/ziglang/zig/pull/10668

edit: I've started work on cleaning this up, see the list above.

### ref
https://github.com/ziglang/zig/issues/10637#issuecomment-1022036688, noting that it might be desirable to remove the LinearFifo from BufferedReader since it had been done for BufferedWriter.

### CODE REVIEW

**Comment by hwchen** [comment]
```
❄️~/zigdev/bench-bufread > ../zig/build/stage3/bin/zig build -Drelease-fast && perf stat zig-out/bin/bench-bufread scratch/test-data.txt

 Performance counter stats for 'zig-out/bin/bench-bufread scratch/test-data.txt':

             20.42 msec task-clock                #    0.989 CPUs utilized          
                 0      context-switches          #    0.000 /sec                   
                 0      cpu-migrations            #    0.000 /sec                   
                10      page-faults               #  489.818 /sec                   
        87,146,865      cycles                    #    4.269 GHz                      (80.42%)
           658,968      stalled-cycles-frontend   #    0.76% frontend cycles idle     (80.41%)
         9,238,396      stalled-cycles-backend    #   10.60% backend cycles idle      (80.40%)
       314,102,036      instructions              #    3.60  insn per cycle         
                                                  #    0.03  stalled cycles per insn  (80.41%)
        91,904,086      branches                  #    4.502 G/sec                    (96.27%)
           570,216      branch-misses             #    0.62% of all branches          (82.08%)

       0.020653192 seconds time elapsed

       0.020717000 seconds user
       0.000000000 seconds sys


❄️~/zigdev/bench-bufread > zig build -Drelease-fast && perf stat zig-out/bin/bench-bufread scratch/test-data.txt

 Performance counter stats for 'zig-out/bin/bench-bufread scratch/test-data.txt':

            144.31 msec task-clock                #    0.997 CPUs utilized          
                 0      context-switches          #    0.000 /sec                   
                 0      cpu-migrations            #    0.000 /sec                   
                12      page-faults               #   83.154 /sec                   
       570,266,676      cycles                    #    3.952 GHz                      (83.37%)
           697,263      stalled-cycles-frontend   #    0.12% frontend cycles idle     (83.37%)
        87,722,281      stalled-cycles-backend    #   15.38% backend cycles idle      (83.37%)
     1,996,966,185      instructions              #    3.50  insn per cycle         
                                                  #    0.04  stalled cycles per insn  (83.37%)
       505,853,940      branches                  #    3.505 G/sec                    (83.37%)
           572,561      branch-misses             #    0.11% of all branches          (83.15%)

       0.144692972 seconds time elapsed

       0.144712000 seconds user
       0.000000000 seconds sys
```

File: lib/std/io/buffered_reader.zig [hunk 1]
```diff
@@ -1,35 +1,39 @@
...
-                    const n = try self.unbuffered_reader.read(writable);
+                    // buf empty, fill it
+                    const n = try self.unbuffered_reader.read(self.buf[0..]);
                     if (n == 0) {
                         // reading from the unbuffered stream returned nothing
                         // so we have nothing left to read.
                         return dest_index;
                     }
-                    self.fifo.update(n);
+                    self.start = 0;
```
Comment (daurnimator) [inline lib/std/io/buffered_reader.zig:33]: This looks the same as the unfleshed out `autoalign` path in `LinearFifo`.
Does performance improve if you try setting that to `true`?

**Review by daurnimator — COMMENTED** [review]
If LinearFifo isn't working well here we probably need to make optimisations over there.
e.g. can we say that in `read` `while (dst_left.len > 0)` is only likely to be true once?

**Comment by andrewrk** [comment]
Thanks @hwchen! Really appreciate the benchmark results - that made it easy to merge.

**Comment by hwchen** [comment]
My pleasure! Thanks for creating a language I enjoy using, and making it easy for me to contribute.

File: lib/std/io/buffered_reader.zig (hunk 1 above)
Comment (hwchen) [reply lib/std/io/buffered_reader.zig:33]: I tried changing to `autoalign=true`but it has about the same (perhaps even a tiny bit worse?) performance.

**Pull request merged**